File: 06_Algorithm/leetcode_2265.go
/**
 * File              : leetcode_2265.go
 * Author            : Andy
 * Date              : 2022.09.07
 * Last Modified Date: 2022.09.07
 * Last Modified By  : Andy
 */

package averageOfSubtree

func dfs(node *TreeNode, ans *int) (int, int) {
	if node == nil {
		return 0, 0
	}

	left_t, left_c := dfs(node.Left)
	right_t, right_c := dfs(node.Right)

	total := left_t + right_t + node.Val
	count := left_c + right_c + 1

	if total/count == node.Val {
		*ans++
	}
	return total, count
}

func averageOfSubtree(root *TreeNode) int {
	ans := 0
	dfs(root, &ans)
	return ans
}

File: Algorithm/leetcode_2265.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# File              : leetcode_2265.py
# Author            : Andy
# Date              : 2022.09.07
# Last Modified Date: 2022.09.07
# Last Modified By  : Andy


class Solution:

    def averageOfSubtree(self, root):
        ans = 0

        def dfs(node):
            nonlocal ans

            if node is None:
                return 0, 0

            left = dfs(node.left)
            right = dfs(node.right)

            total = left[0] + right[0] + node.val
            count = left[1] + right[1] + 1

            if total // count == node.val:
                ans += 1

            return total, count

        dfs(root)
        return ans
